### ISSUE TITLE
intelligent pathFilters for dependencies

### ISSUE THREAD

**Issue by crystalgreen** [dotnet/Nerdbank.GitVersioning#783]
## Problem

I'm using the [pathFilters](https://github.com/dotnet/Nerdbank.GitVersioning/blob/main/doc/pathFilters.md) feature in a .NET _mono repo_. Each project has a separate `version.json` file and produces a nuget package. Some projects depend on each other, so I add those dependencies in _pathFilters_, so the project A version gets bumped if a dependent project B has code changes. This manual configuration in _pathFilters_ is tedious and error prone. 
Sometimes I forget in project A a _pathFilter_ to project B and when a new version of package for B is published (as nuget), it is not updated in another git repo that only references the nuget pkg of A directly.

## Solution (idea)

It would be nice to have an attribute `dependencyFilters` that accepts a _csproj_ file reference. This would lead to the collection of all paths to dependent projects (from `<ProjectReference>`). Each collected path would then be added to _pathFilters_. This could be generalized, i.e. accepting a list of csproj files in case one has one `version.json` for multiple (root) projects (and their dependencies).
Maybe node support (package.json) for in-repo dependencies could make sense as well.

**Comment by AArnott** [comment]
Sorry, we can't do this. It would be a perf killer and no one would use the feature. We have to be able to quickly calculate the full set of path filters *for every single commit in the repo*. But what you're suggesting would require that we MSBuild evaluate the project for every single commit. In fact since lately MSBuild defaults to *transitive project references*, that wouldn't even be enough and ultimately we'd have to evaluate and maybe even execute targets on one or more projects for every single git commit. 

Perhaps you can solve your problem another way though. If your mono repo includes caching of project outputs across builds, it may be possible that you avoid bumping the version of a package that hasn't changed simply by skipping the build of that project (A) entirely. But then project B downstream that consumes it, that *has* changed, would need to know to add a nuget dependency on project A by its older version instead of the version number that would match B's build. This would all be *instead* of using path filters. I've actually implemented this for one of our mono repos at Microsoft, but it's non-trivial and is specific to our internal cloud caching build system.

**Comment by Sam13** [comment]
> This manual configuration in pathFilters is tedious and error prone.
Sometimes I forget in project A a pathFilter to project B and when a new version of package for B is published (as nuget), it is not updated in another git repo that only references the nuget pkg of A directly.

I have a similar problem and was thinking of introducing a check of my _version.json_ files in the CI build outside of NBGV.
Then you have to evaluate your MSBuild dependencies only once.

**Comment by Sam13** [comment]
@crystalgreen
I'm going more into the direction of validating the path filters in the _version.json_ files during build. Then fail the pull request build if they are not correctly set up. To validate them I analyze the project dependencies once.
By doing this NBGV can do it's normal work. Also I do not have to determine the last nuget pkg release somehow (which might be complicated).
I just ensure that NBGV is correctly configured...

**Comment by crystalgreen** [comment]
@Sam13 So you mean to compare path filters against the (transitive) dependencies and ensure they match. The same algorithm could also fix those path filters. Sounds like a good idea. Hopefully a long path filter list won't be a performance issue.

**Comment by crystalgreen** [comment]
Wait. This may give trouble with _SemVer_. Imagine a dependency of our lib gets a bump of mayor or minor version. Our lib only gets a bumped patch version due to git height because of path filters. Due to transitivity, consumers of the lib will also get your dependency (and might use it directly in their code without referencing it explicitly). They implicitly get a potential breaking change though a patch update. Therefore it may be better if our lib bumps also the mayor or minor version instead of only the patch version. See also [this SO post](https://stackoverflow.com/a/53568175/1654527).

Still, the _path filter validation_ idea is better than nothing. Maybe it is worth going one step further and respect SemVer. This may not work with path filters only.

**Comment by AArnott** [comment]
These are good ideas. But in my experience most major version bumps in my dependencies do _not_ result in a major bump in _my_ library because my dependencies are an implementation detail.

**Comment by crystalgreen** [comment]
@AArnott - I guess it depends on ones situation. But in .NET there is no such thing as a private dependency. I actually use implicitly imported libs directly and therefore this may impact me more than you. Maybe my use is bad practice and one should always explicitly pull in the libs one uses directly. This could be enforced by .NET tooling like a compiler warning.

**Comment by crystalgreen** [comment]
Thanks for mentioning PrivateAssets. This seems in deed a way to limit exposure of dependencies with `PrivateAssets=compile` as described [here](https://docs.microsoft.com/en-us/nuget/create-packages/select-assemblies-referenced-by-projects#recommended-one-assembly-per-package). PrivateAssets seem to work for both _PackageReference_ and _ProjectReference_ (see previous link).

I guess, one can still leak types from a dependency if one uses dependency types in arguments or return types of methods of the directly imported lib. If this is true, one could leak breaking changes to the consumer, causing potential compile-time issues. In any case, breaking changes in dependencies could impact the lib and consuming app just by a change in behavior (at runtime). Therefore, dependency updates with breaking changes should be tested thoroughly.

In my case I want to expose dependencies and hence should bump the major version of my lib if one of the exposed dependencies has a major change. Or I rethink this, use `PrivateAssets=compile` and force the lib consumers to pull in the dependencies explicitly.

**Comment by Sam13** [comment]
> Still, the path filter validation idea is better than nothing. Maybe it is worth going one step further and respect SemVer. This may not work with path filters only.

@crystalgreen Agree, having path filters to the project references at least increases the patch version.
I had the problem that NuGet packages have been overwritten but with different dependency requirements.
When going one step further I would compare the dependencies of the "last released" NuGet to the current one to decide which sem ver version part needs to be raised.

>  Or I rethink this, use PrivateAssets=compile and force the lib consumers to pull in the dependencies explicitly.

@crystalgreen If I understand correctly: By using private assets your client application need to provide compatible versions of all your dependencies for all your libraries? Is this feasible when consuming lot's of packages? How do you find compatible versions?

> But in my experience most major version bumps in my dependencies do not result in a major bump in my library because my dependencies are an implementation detail.

@AArnott How do you deal with the situation that different libraries can use the same dependency (cousin dependency). Updating the major version of the dependency in one library might break another library referenced in your client application.

**Comment by crystalgreen** [comment]
@Sam13

> When going one step further I would compare the dependencies of the "last released" NuGet to the current one to decide which sem ver version part needs to be raised.

I was thinking along the same lines. If at least one dependency has major bumped, the lib major version should to be bumped too; same for minor but ignore patch version of dependencies, use local git height as patch.

> By using private assets your client application need to provide compatible versions of all your dependencies for all your libraries?

The _PrivateAsset_ dependencies are still known to nuget, just invisible in Intellisense. I created a test app that references the [Microsoft.ServiceHub.Client version 3.0.3078](https://www.nuget.org/packages/Microsoft.ServiceHub.Client/3.0.3078) package from the [mentioned article](https://docs.microsoft.com/en-us/nuget/create-packages/select-assemblies-referenced-by-projects#recommended-one-assembly-per-package). This pulls in the private dependency [Microsoft.VisualStudio.RemoteControl](https://www.nuget.org/packages/Microsoft.VisualStudio.RemoteControl/) (>= 16.3.32) in version 16.3.32. Then I added a direct reference to that package of version 16.3.44. The VS dependency tree shows that this version is now also used in the private dependency of Microsoft.ServiceHub.Client.
I suppose that nuget uses SemVer and the version constraints of the `PackageReference` to determine if it can find a compatible version for the app that satisfies all constraints. But I haven't yet confirmed this, i.e. with an import of a new major version of such a _cousin_ lib. I'd expect that nuget would fail to import such a conflicting library. Can someone confirm this?

**Comment by AArnott** [comment]
> Or I rethink this, use PrivateAssets=compile and force the lib consumers to pull in the dependencies explicitly.

I would _not_ do that. That defeats the glories of PackageReference allowing you to just pull in what you want and having everything _just work_. I use PrivateAssets=compile for my implementation detail dependencies only, which means I *don't* leak any type references to it in my public API.

> How do you deal with the situation that different libraries can use the same dependency (cousin dependency). Updating the major version of the dependency in one library might break another library referenced in your client application.

I suppose that's a risk. But anytime you have cousin dependencies and they don't *exactly* match, NuGet demands I resolve it through a direct dependency anyway. If such were the case, then the patch update of my library could potentially lead to the consumer realizing they have to make a major version update in this artificially direct dependency, and could then make a decision. So I suppose this means a library should avoid a major update in any of its dependencies (aside from PrivateAssets=all one, which are totally private). But honestly, it hasn't come up as an issue yet.

**Comment by AArnott** [comment]
> I suppose that nuget uses SemVer and the version constraints of the PackageReference to determine if it can find a compatible version for the app that satisfies all constraints. But I haven't yet confirmed this, i.e. with an import of a new major version of such a cousin lib. I'd expect that nuget would fail to import such a conflicting library. Can someone confirm this?

NuGet doesn't make any semver-based version constraint assumptions. If cousin dependencies conflict (that is, they do not _exactly_ match), then nuget issues an error (or maybe a warning) and demands that the top-level project add a direct dependency to specify which version to use. 
Now, nuget also allows version ranges in dependencies (e.g. `[2.0,3.0)`), but I think nuget always resolves this to the lowest matching version within that range. I don't use ranges much myself. Without a range, nuget assumes that any dependency can be filled by that version exactly or *anything* later (even across major versions). With the range, nuget will honor it, so given my `[2.0,3.0)` example, if my library uses that, and its consuming app then tries to directly consume `4.0` of that same dependency, I believe nuget would fail because constraints aren't all satisfied. But if my library had just consumed `2.0`, then the app consuming `4.0` is allowed.

And this is probably the right call, because many libraries are reasonably compatible even across major versions. Most Visual Studio SDK packages are compatible across major versions. VS mostly just revs the major versions of its packages to match the marketing-driven major version of the product itself.

**Comment by crystalgreen** [comment]
@AArnott 

>>    Or I rethink this, use PrivateAssets=compile and force the lib consumers to pull in the dependencies explicitly.

> I would not do that. That defeats the glories of PackageReference allowing you to just pull in what you want and having everything just work.

My test (see above) suggests that private assets _do not_ impact the glories of PackageReference. Nuget is still aware of private package references and updates them if one pulls in an updated lib on top level. The only impact seems to be missing IntelliSense. However, I'm still hesitant of using PrivateAssets as

a) it is an extra effort to configure those dependencies and
b) I want to use those dependencies in the app. Referencing them directly is extra effort. 
c) PrivateAssets just for the sake of controlling import of breaking changes in dependencies doesn't seem the intended purpose.

**Comment by AArnott** [comment]
@crystalgreen By the glories of PackageReference, I'm referring specifically to `PrivateAssets=compile`, when that dependency is used in public API by my library. If someone has a package reference to my library, they should be able to use all the public APIs in my library. But if any of those APIs reference the transitive dependency, then the app using my library will fail to compile when trying to consume those APIs (or overloads of them). It isn't just Intellisense that hides them. The compiler literally isn't given the transitive dependency at all, so overload resolution or calling those APIs on my library directly that the compiler would need a reference to the transitive dependency to resolve will fail. That's a bad experience for my library's consumer. Thus, I never use `PrivateAssets=compile` unless I'm sure my public API won't include types from that dependency.

So for example, let's say I have a library that reads and writes a special 'mysettings.xyz' file. To do so, my library happens to use newtonsoft.json to read/write the file, but the app that uses my library has no business knowing this. They may not even know or care that the file is JSON inside. So in my library, I'll depend on Newtonsoft.Json using `PrivateAssets="compile"` so that when the app uses my library, it doesn't also pollute their intellisense completion lists with a bunch of newtonsoft.json types that they didn't ask for and don't need. If they need newtonsoft.json for their own purposes they can add a direct reference to it. *My* library shouldn't be the avenue they use to acquire newtonsoft and use it directly, because as an implementation detail of my library, I am free to switch it out for System.Text.Json whenever I want, including a patch release, and that would break their app if they needed newtonsoft for other purposes but were only getting it through my library.

I might even justify a major version bump from Newtonsoft v9 to v13 in a patch or minor release of my library, because again, this wouldn't typically break the application that uses my library, unless they are also using newtonsoft.json for other reasons. Now, it is possible this creates a version conflict and a breaking change to the app. But I'd argue that's possible when my library adds a _new_ dependency too. Consider that my library has no Newtonsoft dependency in 1.0. Then in 1.1 I add a Newtonsoft dependency. That new dependency might conflict with an existing dependency the app already has. Does that mean that my library made a breaking change? No, it didn't. It just added a dependency, which is fair game in a minor release, IMO. 

I'd then argue that even if I already had a v9 dependency on Newtonsoft in my 1.0 library, I can 'remove' it and 'add' a v13 dependency in my 1.1 library, if it's a PrivateAssets=compile dependency, because it's an equivalent change. It may or may not be a breaking change to the consuming app based on whether it conflicts with something else they're already doing. But nuget is good at calling these out, and IMO there's more value in my library versioning communicating that it doesn't *directly* have any breaking changes by versioning as 1.1 than to assume the absolute worst and rev the major version every time I add a new dependency or update one of its implementation detail dependencies across a major version.

My two cents.

**Comment by crystalgreen** [comment]
@AArnott 
You are right, `PrivateAssets=compile` also hides the dependency from the compiler (but not from reflection of course). The use case with Newtonsoft you describe is what PrivateAssets was designed for. I agree that one should be able to add/remove (=update) a private dependency during a non-major update. As you say, if the app references a different version of that lib, this creates a version conflict that nuget will detect and call out. This is fine with me as it doesn't go undetected. This is also in line with [SemVer](https://github.com/semver/semver/blob/master/semver.md#what-should-i-do-if-i-update-my-own-dependencies-without-changing-the-public-api).

Back to the discussion of coping with incoming breaking changes via _public_ transitive dependencies.  
These can go undetected if one does a minor version update on the primary lib, assuming minor update means no breaking changes neither from the primary lib nor from public dependencies.
Making those dependencies private with `PrivateAssets=compile` and forcing a consumer to reference them explicitly seems to me abusing the feature and too much effort. Though in principle, I believe it should work, i.e. prevent pulling in unexpected breaking changes. nuget would show the conflict with the old version of the direct reference.  
As already discussed, an alternative solution would be to bump the major version of a lib if one of its public (transitive) dependencies got a major version bump.
Or maybe nuget/VS UI could inform the dev user during a minor update of the primary lib implies a major update of a public dependency. But this would mean that nuget would need to respect SemVer. I previously assumed it would, as the nuget doc [mentions it](https://docs.microsoft.com/en-us/nuget/concepts/package-versioning#semantic-versioning-200) but it does so only in the context of the version format.

**Comment by crystalgreen** [comment]
FYI: I created a Powershell script that adds to the `pathFilters` in a project-local `version.json` all the project reference paths.
See this [gist](https://gist.github.com/crystalgreen/95c258acc475dba9b3c01013393c8e7e)
